File: ruby/lib/html5lib/filters/inject_meta_charset.rb
require 'html5lib/filters/base'

module HTML5lib
  module Filters
    class InjectMetaCharset < Base
      def initialize(source, encoding)
        super(source)
        @encoding = encoding
      end

      def each
        state = :pre_head
        meta_found = @encoding.nil?
        pending = []

        __getobj__.each do |token|
          case token[:type]
          when :StartTag
            state = :in_head if token[:name].downcase == "head"

          when :EmptyTag
            if token[:name].downcase == "meta"
              # replace charset with actual encoding
              token[:data].each_with_index do |(name,value),index|
                if name == 'charset'
                  token[:data][index][1]=@encoding
                  meta_found = true
                end
              end

            elsif token[:name].downcase == "head" and not meta_found
              # insert meta into empty head
              yield(:type => :StartTag, :name => "head", :data => token[:data])
              yield(:type => :EmptyTag, :name => "meta",
                    :data => [["charset", @encoding]])
              yield(:type => :EndTag, :name => "head")
              meta_found = true
              next
            end

          when :EndTag
            if token[:name].downcase == "head" and pending.any?
              # insert meta into head (if necessary) and flush pending queue
              yield pending.shift
              yield(:type => :EmptyTag, :name => "meta",
                    :data => [["charset", @encoding]]) if not meta_found
              yield pending.shift while pending.any?
              meta_found = true
              state = :post_head
            end
          end

          if state == :in_head
            pending << token
          else
            yield token
          end
        end
      end
    end
  end
end

File: python/src/filters/inject_meta_charset.py
import _base

class Filter(_base.Filter):
    def __init__(self, source, encoding):
        _base.Filter.__init__(self, source)
        self.encoding = encoding

    def __iter__(self):
        state = "pre_head"
        meta_found = (self.encoding is None)
        pending = []

        for token in _base.Filter.__iter__(self):
            type = token["type"]
            if type == "StartTag":
                if token["name"].lower() == "head":
                    state = "in_head"

            elif type == "EmptyTag":
                if token["name"].lower() == "meta":
                   # replace charset with actual encoding
                   for i,(name,value) in enumerate(token["data"]):
                       if name == 'charset':
                          token["data"][i] = (token["data"][i][0], self.encoding)
                          meta_found = True

                elif token["name"].lower() == "head" and not meta_found:
                    # insert meta into empty head
                    yield {"type": "StartTag", "name": "head",
                           "data": token["data"]}
                    yield {"type": "EmptyTag", "name": "meta",
                           "data": [["charset", self.encoding]]}
                    yield {"type": "EndTag", "name": "head"}
                    meta_found = True
                    continue

            elif type == "EndTag":
                if token["name"].lower() == "head" and pending:
                    # insert meta into head (if necessary) and flush pending queue
                    yield pending.pop(0)
                    if not meta_found:
                        yield {"type": "EmptyTag", "name": "meta",
                               "data": [["charset", self.encoding]]}
                    while pending:
                        yield pending.pop(0)
                    meta_found = True
                    state = "post_head"

            if state == "in_head":
                pending.append(token)
            else:
                yield token
